src/dpg_slam/dpg_slam.cc

```diff
@@ -676,25 +738,64 @@ namespace dpg_slam {
+             if ((occ_grid_entry.second == OCCUPIED) && (local_submap_occupancy == FREE)) {
+                   std::vector<PointIdInfo> new_added = current_node_occ_grid.getPointsInOccCell(occ_grid_entry.first);
+                   added_points_at_current_node.insert(added_points_at_current_node.end(), new_added.begin(), new_added.end());
+		   ROS_INFO_STREAM("Added");
+            	} else if ((occ_grid_entry.second == FREE) && (local_submap_occupancy == OCCUPIED)) {
+                   std::vector<PointIdInfo> new_removed = local_submap_occ_grid.getPointsInOccCell(occ_grid_entry.first);
+                   removed_points_at_current_node.insert(removed_points_at_current_node.end(), new_removed.begin(), new_removed.end());
+		   ROS_INFO_STREAM("Removed");
+            	}
             }
-        }
+ 	 
+	 // if newly detected changes at node are non zero, check the score.
+	 if (added_points_at_current_node.size()+removed_points_at_current_node.size() > 0) {
+	 	commitChanges = computeBinScoreAndCommitLabelsForNode(added_points_at_current_node,
+									   removed_points_at_current_node);
+	 }
+	 if (!commitChanges) {
+		return;
+	 }
+
+	 // commit changes if we get here.
+	 added_points.insert(added_points.end(), added_points_at_current_node.begin(), added_points_at_current_node.end());
+	 removed_points.insert(removed_points.end(), removed_points_at_current_node.begin(), removed_points_at_current_node.end());
     }
 
-    std::vector<DpgNode> DpgSLAM::updateActiveAndDynamicMaps(const std::vector<DpgNode> &nodes, const std::vector<dpgMapPoint> &removedPoints){
-
-        std::vector<DpgNode> inactiveNodes;
-        //TODO: Implement Alg. 2 from the paper.
-
-
-        return inactiveNodes;
+    bool DpgSLAM::computeBinScoreAndCommitLabelsForNode(const std::vector<PointIdInfo> &added_points,
+		    					const std::vector<PointIdInfo> &removed_points) {
+
+         uint16_t totalBins = dpg_parameters_.num_bins_for_change_detection_;
+	 double change_threshold = dpg_parameters_.delta_change_threshold_;
+	 std::unordered_set<uint16_t> changedBins;
+	 bool commitChanges = false;
+	 double current_changed_ratio = 0.0;
+	 std::vector<PointIdInfo> changedPoints = added_points;
+	 changedPoints.insert(changedPoints.end(), removed_points.begin(), removed_points.end());
+
+	 for(const PointIdInfo& point : changedPoints) {
+		DpgNode node = dpg_nodes_[point.node_num_];
+                Measurement MeasurementOfPoint = node.getMeasurement();
+		std::vector<MeasurementPoint> ranges = MeasurementOfPoint.getMeasurements();
+		std::pair<float, float> angleRanges = MeasurementOfPoint.getMeasurementRange();
+                float angle = ranges[point.point_num_in_node_].getAngle();
+		uint32_t binNumber = round(totalBins*(angle-angleRanges.first)/(angleRanges.second-angleRanges.first));
+		changedBins.insert(binNumber);
+		current_changed_ratio = changedBins.size()/totalBins;
```

Comment:
We need to get the angle relative to the current node rather than the containing node. This should instead be something like this (definitely double check this, I might've transposed some variables or gotten inverseTransform/transform mixed up). 

```
DpgNode curr_pose_chain_node; // TODO need to get this from the added points or pass it in
float angle_min = curr_pose_chain_node.getMeasurementRange().first;
float angle_max = curr_pose_chain_node.getMeasurementRange().second;
float bin_increment = (angle_max - angle_min) / totalBins;
std::pair<Vector2f, float> curr_lidar_in_map = transform(lidar_in_curr_node.first, lidar_in_curr_node.second)
for (const PointIdInfo &point : changedPoints) {
    Vector2f point_loc_map = // Get point location in map frame;
    Vector2f point_loc_rel_lidar = inverseTransform(point_loc_map, 0, curr_lidar_in_map.first, curr_lidar_in_map.second);
    float angle_rel_curr_node = atan2(point_loc_rel_lidar.y(), point_loc_rel_lidar.x());
    if ((angle_rel_curr_node > angle_max) || (angle_rel_curr_node < angle_min)) {
        continue;
    }
    uint16_t bin_num = (angle_rel_curr_node - angle_min) / bin_increment;
    // Rest of the stuff you have
}

```